model/stats/db.go

```diff
@@ -442,15 +443,309 @@ func statsToUpdatePipeline(projectId string, start time.Time, end time.Time) []b
+	case GroupByDistro:
+		fields = []string{"_id.date", "_id.variant", "_id.task_name", "_id.test_file", "_id.distro"}
+		operators = []string{dateOperator, "$gt", "$gt", "$gt", "$gt"}
+		values = []interface{}{filter.StartAt.Date, filter.StartAt.BuildVariant, filter.StartAt.Task, filter.StartAt.Test, filter.StartAt.Distro}
+	}
+
+	matchExpr["$or"] = buildPaginationOrBranches(fields, operators, values)
+}
+
+// Creates an aggregation pipeline to query task statistics.
+func taskStatsQueryPipeline(filter *StatsFilter) []bson.M {
+	matchExpr := buildMatchStageForTask(filter)
+
+	return []bson.M{
+		matchExpr,
+		buildAddFieldsDateStage("date", filter.AfterDate, filter.BeforeDate, filter.GroupNumDays),
+		{"$group": bson.M{
+			"_id":                    buildGroupId(filter.GroupBy),
+			"num_success":            bson.M{"$sum": "$num_success"},
+			"num_failed":             bson.M{"$sum": "$num_failed"},
+			"num_timeout":            bson.M{"$sum": "$num_timeout"},
+			"num_test_failed":        bson.M{"$sum": "$num_test_failed"},
+			"num_system_failed":      bson.M{"$sum": "$num_system_failed"},
+			"num_setup_failed":       bson.M{"$sum": "$num_setup_failed"},
+			"total_duration_success": bson.M{"$sum": bson.M{"$multiply": array{"$num_success", "$avg_duration_success"}}},
+		}},
+		{"$project": bson.M{
+			"task_name":         "$_id.task_name",
+			"variant":           "$_id.variant",
+			"distro":            "$_id.distro",
+			"date":              "$_id.date",
+			"num_success":       1,
+			"num_failed":        1,
+			"num_total":         bson.M{"$add": array{"$num_success", "$num_failed"}},
+			"num_timeout":       1,
+			"num_test_failed":   1,
+			"num_system_failed": 1,
+			"num_setup_failed":  1,
+			"avg_duration_success": bson.M{"$cond": bson.M{"if": bson.M{"$ne": array{"$num_success", 0}},
+				"then": bson.M{"$divide": array{"$total_duration_success", "$num_success"}},
+				"else": nil}},
+		}},
+		{"$sort": bson.D{
+			{Name: "date", Value: sortDateOrder(filter.Sort)},
+			{Name: "variant", Value: 1},
+			{Name: "task_name", Value: 1},
+			{Name: "distro", Value: 1},
+		}},
+		{"$limit": filter.Limit},
+	}
+}
+
+// Builds the match stage of the task query pipeline based on the filter options.
+func buildMatchStageForTask(filter *StatsFilter) bson.M {
+	match := bson.M{
+		"_id.date": bson.M{
+			"$gte": filter.AfterDate,
+			"$lt":  filter.BeforeDate,
+		},
+		"_id.project":   filter.Project,
+		"_id.requester": bson.M{"$in": filter.Requesters},
+	}
+	addMatchIn(match, "_id.task_name", filter.Tasks)
+	addMatchIn(match, "_id.variant", filter.BuildVariants)
+	addMatchIn(match, "_id.distro", filter.Distros)
+
+	if filter.StartAt != nil {
+		addMatchTaskPagination(match, filter)
+	}
+
+	return bson.M{"$match": match}
+}
+
+// Adds to an existing $match expression the conditions imposed by the filter StartAt field
+func addMatchTaskPagination(matchExpr bson.M, filter *StatsFilter) {
```

Comment:
can we have these functions return something, rather than mutating some internal state. 

It's not *always* true, but I think its often true that functions that don't return anything should probably be methods on something (like filter?) and everything else should return things even if you end up having do do something like `query = modifyQuery(query)` at the call site, it makes for a much easier to use api.

Reply:
Done.